server/util.go

```diff
@@ -280,13 +286,77 @@ func dumpBinaryRow(buffer []byte, columns []*ColumnInfo, row chunk.Row) ([]byte,
 	return buffer, nil
 }
 
-func dumpTextRow(buffer []byte, columns []*ColumnInfo, row chunk.Row) ([]byte, error) {
+type textDumper struct {
+	encoding charset.Encoding
+	buffer   []byte
+	isBinary bool
+	isNull   bool
+}
+
+// newTextDumper reads the system variable @@character_set_results and creates the textDumper.
+func newTextDumper(ctx context.Context, sysVars *variable.SessionVars) *textDumper {
+	chs, err := variable.GetSessionOrGlobalSystemVar(sysVars, variable.CharacterSetResults)
+	if err != nil {
+		logutil.Logger(ctx).Info("get character_set_results system variable failed", zap.Error(err))
+	}
+	chsName := charset.Format(chs)
+	ret := &textDumper{}
+	ret.encoding.UpdateEncoding(chsName)
+	ret.isBinary = chsName == charset.CharsetBinary
+	ret.isNull = len(chsName) == 0
+	return ret
+}
+
+// clean prevent the textDumper from holding too much memory.
+func (d *textDumper) clean() {
+	d.buffer = nil
+}
+
+func (d *textDumper) resetCharset() func() {
+	origin := d.encoding.Name()
+	return func() {
+		if d.isBinary || d.isNull {
+			d.encoding.UpdateEncoding(charset.Formatted(origin))
+		}
+	}
+}
+
+func (d *textDumper) updateCharset(chsID uint16) {
+	if d.isNull || d.isBinary {
```

Comment:
only support update charset for @@character_set_results="" or Binary charset? 
better add a comment here